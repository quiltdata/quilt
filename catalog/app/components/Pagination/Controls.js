import PT from 'prop-types';
import React from 'react';
import {
  compose,
  setDisplayName,
  setPropTypes,
  withProps,
} from 'recompose';
import styled from 'styled-components';
import { FormattedMessage as FM } from 'react-intl';
import IconButton from 'material-ui/IconButton';

import MIcon from 'components/MIcon';

import messages from './messages';

const UNIT_SIZE = '24px';

const Container = styled.div`
  display: flex;
  margin-top: 16px;
`;

const Chevron = compose(
  setPropTypes({
    direction: PT.oneOf(['left', 'right']).isRequired,
  }),
  withProps(({ direction }) => ({
    children: <MIcon>{`chevron_${direction}`}</MIcon>,
    style: {
      height: UNIT_SIZE,
      padding: 0,
      width: UNIT_SIZE,
    },
  })),
)(IconButton);

const Pages = styled.span`
  margin-left: 12px;
`;

export default compose(
  setPropTypes({
    page: PT.number.isRequired,
    pages: PT.number.isRequired,
    nextPage: PT.func.isRequired,
    prevPage: PT.func.isRequired,
  }),
  setDisplayName('Pagination.Controls'),
// eslint-disable-next-line object-curly-newline
)(({ page, pages, nextPage, prevPage }) => pages <= 1 ? null : (
  <Container>
    <Chevron
      direction="left"
      onClick={prevPage}
      disabled={page <= 1}
    />
    <Chevron
      direction="right"
      onClick={nextPage}
      disabled={page >= pages}
    />
    <Pages>{page} <FM {...messages.of} /> {pages}</Pages>
  </Container>
));
