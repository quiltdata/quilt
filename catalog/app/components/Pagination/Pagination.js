import PT from 'prop-types'
import * as R from 'ramda'
import * as React from 'react'
import * as RC from 'recompose'

import * as RT from 'utils/reactTools'
import usePrevious from 'utils/usePrevious'

const PER_PAGE = 10

const useGetter = (value, get) =>
  React.useMemo(() => (value == null ? value : get(value)), [value, get])

function useHasChanged(value, getKey = R.identity) {
  const key = useGetter(value, getKey)
  const oldValue = usePrevious(value)
  const oldKey = useGetter(oldValue, getKey)
  if (R.is(Array, key) && R.is(Array, oldKey)) {
    // if new list == old list + more items (appended), consider it unchanged
    // to avoid resetting pagination on adding more items to the paginated set
    return !R.startsWith(oldKey, key)
  }
  return !R.equals(key, oldKey)
}

export const usePagination = (
  items,
  { getItemId = R.identity, perPage: initialPerPage = PER_PAGE, onChange } = {},
) => {
  const [page, setPage] = React.useState(1)
  const [perPage, setPerPage] = React.useState(initialPerPage)

  const pages = Math.max(1, Math.ceil(items.length / perPage))

  const goToPage = React.useMemo(() => R.pipe(R.clamp(1, pages), setPage), [
    pages,
    setPage,
  ])

  const nextPage = React.useCallback(() => goToPage(page + 1), [goToPage, page])

  const prevPage = React.useCallback(() => goToPage(page - 1), [goToPage, page])

  const getKey = useGetter(getItemId, R.map)
  if (useHasChanged(items, getKey) && page !== 1) {
    // reset to page 1 if items change (but not if appended)
    goToPage(1)
  }

  const offset = (page - 1) * perPage

  const paginate = React.useMemo(() => R.slice(offset, offset + perPage), [
    offset,
    perPage,
  ])
  const paginated = useGetter(items, paginate)

  usePrevious(perPage, (prev) => {
    if (prev && perPage !== prev) {
      goToPage(Math.floor(((page - 1) * prev) / perPage) + 1)
    }
  })

  usePrevious(page, (prev) => {
    if (page !== prev && onChange) onChange(prev, page)
  })

  return {
    paginated,
    total: items.length,
    from: offset + 1,
    to: offset + perPage,
    perPage,
    setPerPage,
    page,
    pages,
    nextPage,
    prevPage,
    goToPage,
  }
}

export const use = usePagination

export const Paginate = RT.composeComponent(
  'Pagination.Paginate',
  RC.setPropTypes({
    items: PT.array.isRequired,
    children: PT.func.isRequired,
  }),
  ({ items, children, ...props }) => children(use(items, props)),
)
