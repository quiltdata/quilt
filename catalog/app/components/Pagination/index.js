import PT from 'prop-types';
import React from 'react';
import {
  compose,
  lifecycle,
  pure,
  setDisplayName,
  setPropTypes,
  withStateHandlers,
  withProps,
} from 'recompose';
import styled from 'styled-components';
import { FormattedMessage as FM } from 'react-intl';
import IconButton from 'material-ui/IconButton';

import MIcon from 'components/MIcon';

import messages from './messages';

const PER_PAGE = 10;
const UNIT_SIZE = '24px';

const Controls = styled.div`
  display: flex;
  margin-top: 16px;
`;

const Chevron = compose(
  setPropTypes({
    direction: PT.oneOf(['left', 'right']).isRequired,
  }),
  withProps(({ direction }) => ({
    children: <MIcon>{`chevron_${direction}`}</MIcon>,
    style: {
      height: UNIT_SIZE,
      padding: 0,
      width: UNIT_SIZE,
    },
  })),
)(IconButton);

const Pages = styled.span`
  margin-left: 12px;
`;

export default compose(
  setPropTypes({
    items: PT.array.isRequired,
    children: PT.func.isRequired,
  }),
  withProps(({ items }) => ({
    pages: Math.max(1, Math.ceil(items.length / PER_PAGE)),
  })),
  withStateHandlers(
    { page: 1 },
    {
      nextPage: ({ page }, { pages }) => () => ({ page: Math.min(pages, page + 1) }),
      prevPage: ({ page }) => () => ({ page: Math.max(1, page - 1) }),
      resetPage: () => () => ({ page: 1 }),
    },
  ),
  pure,
  lifecycle({
    componentWillReceiveProps({ items }) {
      if (items !== this.props.items) this.props.resetPage();
    },
  }),
  withProps(({ page, items }) => {
    const offset = (page - 1) * PER_PAGE;
    return {
      items: items.slice(offset, offset + PER_PAGE),
    };
  }),
  setDisplayName('Pagination'),
)(({
  items,
  page,
  pages,
  nextPage,
  prevPage,
  children,
}) => (
  <div>
    {children({ items })}
    {!!items.length && pages > 1 &&
      <Controls>
        <Chevron
          direction="left"
          onClick={prevPage}
          disabled={page <= 1}
        />
        <Chevron
          direction="right"
          onClick={nextPage}
          disabled={page >= pages}
        />
        <Pages>{page} <FM {...messages.of} /> {pages}</Pages>
      </Controls>
    }
  </div>
));
