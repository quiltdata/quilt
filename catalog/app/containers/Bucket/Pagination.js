import cx from 'classnames'
import * as R from 'ramda'
import * as React from 'react'
import { Link } from 'react-router-dom'
import * as M from '@material-ui/core'

const Gap = Symbol('gap')

const MAX_PAGE_BUTTONS = 8 // fits on 320px-wide screen

// 1-based
// total | current  | output
// ------|----------|----------------------
// 8     | *        | [1, 2, 3, 4, 5, 6, 7, 8]
// 100   | 1...4    | [1, 2, 3, 4, 5, 6, <G>, 100]
// 100   | 5...95   | [1, <G>, N-1, N, N+1, N+2, <G>, 100]
// 100   | 96...100 | [1, <G>, 95, 96, 97, 98, 99, 100]
const displayRange = (total, current) => {
  if (total <= MAX_PAGE_BUTTONS) return R.range(1, total + 1)
  if (current <= MAX_PAGE_BUTTONS - 4) {
    return [...R.range(1, MAX_PAGE_BUTTONS - 1), Gap, total]
  }
  if (current >= total - MAX_PAGE_BUTTONS + 4) {
    return [1, Gap, ...R.range(total - MAX_PAGE_BUTTONS + 3, total + 1)]
  }
  return [1, Gap, ...R.range(current - 1, current + 3), Gap, total]
}

const usePaginationStyles = M.makeStyles((t) => ({
  button: {
    background: t.palette.common.white,
    paddingLeft: t.spacing(1.5),
    paddingRight: t.spacing(1.5),
  },
  current: {
    color: t.palette.primary.contrastText,
    backgroundColor: t.palette.primary.main,
    '&:hover': {
      backgroundColor: t.palette.primary.dark,
    },
  },
  gap: {
    cursor: 'default',
    pointerEvents: 'none',
  },
}))

export default function Pagination({ page, pages, makePageUrl }) {
  const classes = usePaginationStyles()
  const range = React.useMemo(() => displayRange(pages, page), [pages, page])

  return (
    <M.Box display="flex" justifyContent="center" mt={3} mb={{ xs: 5, sm: 0 }}>
      <M.ButtonGroup variant="contained">
        {range.map((p, i) =>
          p === Gap ? (
            <M.Button
              // eslint-disable-next-line react/no-array-index-key
              key={`gap:${i}`}
              className={cx(classes.button, classes.gap)}
              component="span"
            >
              &hellip;
            </M.Button>
          ) : (
            <M.Button
              key={`page:${p}`}
              component={Link}
              className={cx(classes.button, { [classes.current]: page === p })}
              to={makePageUrl(p)}
            >
              {p}
            </M.Button>
          ),
        )}
      </M.ButtonGroup>
    </M.Box>
  )
}
