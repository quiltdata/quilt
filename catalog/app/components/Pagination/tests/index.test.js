import PT from 'prop-types';
import React from 'react';
import { shallow } from 'enzyme';

import Pagination from '../index';


describe('<Pagination />', () => {
  const itemsEmpty = [];
  const items1page = [
    { id: 1 },
    { id: 2 },
    { id: 3 },
  ];
  /*
  const items2pages = [
    { id: 1 },
    { id: 2 },
    { id: 3 },
    { id: 4 },
    { id: 5 },
    { id: 6 },
    { id: 7 },
    { id: 8 },
    { id: 9 },
    { id: 10 },
    { id: 11 },
    { id: 12 },
  ];
  */
  // TODO: import PER_PAGE constant?
  // const items2pagesP1 = items2pages.slice(0, 10);
  // const items2pagesP2 = items2pages.slice(10, 20);
  // const items2pagesChanged = [
  //   { id: 2 },
  //   { id: 3 },
  //   { id: 4 },
  //   { id: 5 },
  //   { id: 6 },
  //   { id: 7 },
  //   { id: 8 },
  //   { id: 9 },
  //   { id: 10 },
  //   { id: 11 },
  //   { id: 12 },
  // ];

  const SimpleList = ({ items }) =>
    <list>{items.map(({ id }) => <item key={id}>{id}</item>)}</list>;

  SimpleList.propTypes = {
    items: PT.array.isRequired,
  };

  const render = (items, children = SimpleList) =>
    shallow(<Pagination items={items}>{children}</Pagination>)
      // dive into HoCs
      .dive().dive().dive().dive().dive().dive().dive().dive().dive().dive().dive().dive().dive();

  // TODO
  // const clickNext = (wrapper) =>
  //   wrapper.find('[direction="right"]').simulate('click');
  // const clickPrev = (wrapper) =>
  //   wrapper.find('[direction="left"]').simulate('click');

  describe('when there is 0 items', () => {
    it('should match the snapshot', () => {
      expect(render(itemsEmpty).render()).toMatchSnapshot();
    });

    it('should pass items to the children', () => {
      const children = jest.fn();
      render(itemsEmpty, children);
      expect(children).toBeCalledWith({ items: itemsEmpty });
    });
  });

  describe('when there are items for only one page', () => {
    it('should match the snapshot', () => {
      expect(render(items1page).render()).toMatchSnapshot();
    });

    it('should pass items to the children', () => {
      const children = jest.fn();
      render(items1page, children);
      expect(children).toBeCalledWith({ items: items1page });
    });
  });

  /* broken bc PER_PAGE is set to larger value
  describe('when there are items for more than one page', () => {
    it('should match the snapshot', () => {
      expect(render(items2pages)).toMatchSnapshot();
    });

    it('should pass paginated items to the children', () => {
      const children = jest.fn();
      render(items2pages, children);
      expect(children).toBeCalledWith({ items: items2pagesP1 });
    });
  });
  */

  // TODO
  // describe('when next button is clicked', () => {
  //   describe('and there is next page', () => {
  //     it('should go to the next page', () => {
  //     });
  //   });
  //   describe('and there is no next page', () => {
  //   });
  // });

  // describe('when back button is clicked', () => {
  //   describe('and there is previous page', () => {
  //   });
  //   describe('and there is no previous page', () => {
  //   });
  // });

  // describe('when the items prop gets changed', () => {
  //   it('should reset the page number', () => {
  //   });
  // });
});
