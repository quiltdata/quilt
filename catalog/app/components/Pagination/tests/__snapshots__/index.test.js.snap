// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> when there are items for only one page should match the snapshot 1`] = `
<list>
  <item>
    1
  </item>
  <item>
    2
  </item>
  <item>
    3
  </item>
</list>
`;

exports[`<Pagination /> when there is 0 items should match the snapshot 1`] = `<list />`;
